homeassistant/components/isy994/services.py

```diff
@@ -0,0 +1,408 @@
+            vol.Exclusive(CONF_ADDRESS, SCHEMA_GROUP): cv.string,
+            vol.Required(CONF_COMMAND): vol.In(VALID_PROGRAM_COMMANDS),
+            vol.Optional(CONF_ISY): cv.string,
+        }
+    ),
+)
+
+SERVICE_RUN_NETWORK_RESOURCE_SCHEMA = vol.All(
+    cv.has_at_least_one_key(CONF_ADDRESS, CONF_NAME),
+    vol.Schema(
+        {
+            vol.Exclusive(CONF_NAME, SCHEMA_GROUP): cv.string,
+            vol.Exclusive(CONF_ADDRESS, SCHEMA_GROUP): vol.Coerce(int),
+            vol.Optional(CONF_ISY): cv.string,
+        }
+    ),
+)
+
+
+@callback
+def async_setup_services(hass: HomeAssistantType):
+    """Create and register services for the ISY integration."""
+    existing_services = hass.services.async_services().get(DOMAIN)
+    if existing_services and any(
+        service in INTEGRATION_SERVICES for service in existing_services.keys()
+    ):
+        # Integration-level services have already been added. Return.
+        return
+
+    async def async_system_query_service_handler(service):
+        """Handle a system query service call."""
+        address = service.data.get(CONF_ADDRESS)
+        isy_name = service.data.get(CONF_ISY)
+
+        for config_entry_id in hass.data[DOMAIN]:
+            isy = hass.data[DOMAIN][config_entry_id][ISY994_ISY]
+            if isy_name and not isy_name == isy.configuration["name"]:
+                continue
+            # If an address is provided, make sure we query the correct ISY.
+            # Otherwise, query the whole system on all ISY's connected.
+            if address and isy.nodes.get_by_id(address) is not None:
+                _LOGGER.debug(
+                    "Requesting query of device %s on ISY %s",
+                    address,
+                    isy.configuration["uuid"],
+                )
+                hass.async_add_executor_job(isy.query, address)
+                return
+            _LOGGER.debug(
+                "Requesting system query of ISY %s", isy.configuration["uuid"]
+            )
+            hass.async_add_executor_job(isy.query)
+
+    async def async_run_network_resource_service_handler(service):
+        """Handle a network resource service call."""
+        address = service.data.get(CONF_ADDRESS)
+        name = service.data.get(CONF_NAME)
+        isy_name = service.data.get(CONF_ISY)
+
+        for config_entry_id in hass.data[DOMAIN]:
+            isy = hass.data[DOMAIN][config_entry_id][ISY994_ISY]
+            if isy_name and not isy_name == isy.configuration["name"]:
+                continue
+            if not hasattr(isy, "networking") or isy.networking is None:
+                continue
+            command = None
+            if address:
+                command = isy.networking.get_by_id(address)
+            if name:
+                command = isy.networking.get_by_name(name)
+            if command is not None:
+                hass.async_add_executor_job(command.run)
```

Comment:
We should probably await this.

Reply:
Do all of these executor calls need to be awaited (e.g. line 168, 191 above, 234, 259 below)?

For reference, the underlying call is a `requests`' GET command on a REST endpoint, results are logged as success or failure only.

I will test again, I think I had an await at some point and got an I/O warning.

Reply:
> Do all of these executor calls need to be awaited (e.g. line 168, 191 above, 234, 259 below)?

Yes
> 
> For reference, the underlying call is a `requests`' GET command on a REST endpoint, results are logged as success or failure only.
> 
> I will test again, I think I had an await at some point and got an I/O warning.

As long as they are wrapped in the `async_add_executor_job` you should be good to go.